yi-core/src/Yi/Clip.hs

```diff
@@ -0,0 +1,46 @@
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+-- |
+-- Module      :  Yi.Clip
+-- License     :  GPL-2
+-- Maintainer  :  yi-devel@googlegroups.com
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- A proxy of clipboard.
+
+module Yi.Clip ( setClipboard
+               , getClipboard
+               )
+    where
+
+import qualified System.Hclip        as H (getClipboard, setClipboard)
+import           Data.IORef
+import           System.IO.Unsafe
+
+import           Yi.Types            (configDisableSystemClipboard, askCfg)
+import           Yi.Utils            (io)
+import           Yi.Keymap           (YiM)
+
+clipboard :: IORef String
+clipboard = unsafePerformIO $ newIORef ""
```

Comment:
1. Don't store String, use YiRope straight away.
2. Get rid of this global IORef completely, what's the point of it? YiM can carry the state you need. If you really want an IORef, put the IORef in YiM and update it that way. `getClipboard` can then determine where to read clipboard state from.

Reply:
1. We do not need performance. Clipboard is poor one. 
    When vty-mode, Hclip calls a command of ```xclip```, it is really slow.
 　So I think we do not need YiRope.
2.  I know YiM can carry the clipboard data. 
     I just do not want to increase the mutable state of YiM.
     If you do not want to use IORef by all means, I'll use the state of YiM.

Reply:
>1. We do not need performance. Clipboard is poor one.

There's zero reason to throw away performance. In this case it's more about convenience of any users: why force them to convert from string all the time? Further, String sucks for _space_, there's no reason to hold onto the clipboard contents as String and put completely unnecessary pressure on GHC's copying GC.

>2. I just do not want to increase the mutable state of YiM.

Just add it to YiM state, it's a lot easier to reason about than some random floating globals.

Reply:
Hi @noughtmare ,
Do you have any comments?

Reply:
Please excuse me for my late reply, I was very busy with work for my university.

I believe that quality is very important, especially when an objectively better solution is known. I understand that it is not always possible, a balance needs to be struck between effort and quality.

I will explain my interpretation of @Fuuzetsu's proposal in more detail. The proposal consists of two parts. Firstly, to have the virtual clipboard provide and store `YiString`s, to prevent conversion between `YiString`s and `String`s which saves memory and CPU time. The system clipboard can then has to explicitly convert the `YiString`s to `String`s, because the system clipboard can't store `String`s. The second part is to use the dynamic state inside the `YiM` monad to store the virtual clipboard contents, instead of a loose reference.

I think the first part of the proposal is objectively an improvement. We don't expect performance from the clipboard, but we do expect it from the editor as a whole and many small performance deficiencies can add up.

The second part, however, is not so clear-cut. The loose reference is not exposed by the `Yi.Clip` module, so it will not be accessible to other parts of the editor. Storing the clipboard contents in the `YiM` monad *will* make it accessible to the whole editor, to me *that* sounds harder to reason about.

@Fuuzetsu and @junjihashimoto have I convinced you or do you still have objections?

Reply:
>YiString

Agree, use it.

>Storing the clipboard contents in the YiM monad will make it accessible to the whole editor, to me that sounds harder to reason about.

This is not a Bad Thing™. If anything, this allows you to make `getClipboard` and `setClipboard` run in `YiM` instead of `IO`: this means you can actually stop someone using these somewhere arbitrary and things changing uncontrollably. 

It's a bit like using global IORef instead of ReaderT with IORef in the env… I don't like it.

Having said that, this is about emulating system clipboard which is arguably this global system thing. So I don't care _too_ strongly about it. If you want it in IORefs, hide them in the module and I'll stomach it.